File: release/com/example/budgettingapp/AddExpenseActivity.java
package com.example.budgettingapp;

import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.Intent;
import android.os.Bundle;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import com.example.budgettingapp.data.DataManager;
import com.example.budgettingapp.data.Expense;
import com.example.budgettingapp.databinding.ActivityAddExpenseBinding;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.*;

@kotlin.Metadata(mv = {1, 9, 0}, k = 1, xi = 48, d1 = {"\u0000X\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\b\u0005\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000b\n\u0002\b\b\u0018\u00002\u00020\u0001B\u0005\u00a2\u0006\u0002\u0010\u0002J\u0012\u0010\r\u001a\u00020\u000e2\b\u0010\u000f\u001a\u0004\u0018\u00010\u0010H\u0014J\u0010\u0010\u0011\u001a\u00020\u00122\u0006\u0010\u0013\u001a\u00020\u0014H\u0002J\u0018\u0010\u0015\u001a\u00020\u00042\u0006\u0010\u0013\u001a\u00020\u00142\u0006\u0010\u0016\u001a\u00020\u0014H\u0002J\u0010\u0010\u0017\u001a\u00020\u00182\u0006\u0010\u0016\u001a\u00020\u0014H\u0002J\b\u0010\u0019\u001a\u00020\u000eH\u0002J\b\u0010\u001a\u001a\u00020\u000eH\u0002J\b\u0010\u001b\u001a\u00020\u000eH\u0002J\u0010\u0010\u001c\u001a\u00020\u000e2\u0006\u0010\u001d\u001a\u00020\u001eH\u0002J8\u0010\u001f\u001a\u00020 2\u0006\u0010!\u001a\u00020\u00142\u0006\u0010\"\u001a\u00020\u00142\u0006\u0010#\u001a\u00020\u00142\u0006\u0010$\u001a\u00020\u00142\u0006\u0010%\u001a\u00020\u00142\u0006\u0010&\u001a\u00020\u0014H\u0002J\u0018\u0010\'\u001a\u00020 2\u0006\u0010\"\u001a\u00020\u00042\u0006\u0010#\u001a\u00020\u0004H\u0002R\u0016\u0010\u0003\u001a\n \u0005*\u0004\u0018\u00010\u00040\u0004X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0016\u0010\u0006\u001a\n \u0005*\u0004\u0018\u00010\u00070\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0016\u0010\b\u001a\n \u0005*\u0004\u0018\u00010\u00070\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\t\u001a\u00020\nX\u0082.\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u000b\u001a\u00020\fX\u0082.\u00a2\u0006\u0002\n\u0000\u00a8\u0006("}, d2 = {"Lcom/example/budgettingapp/AddExpenseActivity;", "Landroidx/appcompat/app/AppCompatActivity;", "()V", "CURRENT_DATE", "Ljava/time/LocalDateTime;", "kotlin.jvm.PlatformType", "DATE_FORMATTER", "Ljava/time/format/DateTimeFormatter;", "TIME_FORMATTER", "binding", "Lcom/example/budgettingapp/databinding/ActivityAddExpenseBinding;", "dataManager", "Lcom/example/budgettingapp/data/DataManager;", "onCreate", "", "savedInstanceState", "Landroid/os/Bundle;", "parseDate", "Ljava/time/LocalDate;", "dateString", "", "parseDateTime", "timeString", "parseTime", "Ljava/time/LocalTime;", "saveExpense", "setupViews", "showDatePicker", "showTimePicker", "timeEditText", "Landroid/widget/EditText;", "validateInput", "", "date", "startTime", "endTime", "description", "category", "amountText", "validateTimeRange", "app_release"})
public final class AddExpenseActivity extends androidx.appcompat.app.AppCompatActivity {
    private com.example.budgettingapp.databinding.ActivityAddExpenseBinding binding;
    private com.example.budgettingapp.data.DataManager dataManager;
    private final java.time.format.DateTimeFormatter DATE_FORMATTER = null;
    private final java.time.format.DateTimeFormatter TIME_FORMATTER = null;
    private final java.time.LocalDateTime CURRENT_DATE = null;
    
    public AddExpenseActivity() {
        super();
    }
    
    @java.lang.Override
    protected void onCreate(@org.jetbrains.annotations.Nullable
    android.os.Bundle savedInstanceState) {
    }
    
    private final void setupViews() {
    }
    
    private final void showDatePicker() {
    }
    
    private final void showTimePicker(android.widget.EditText timeEditText) {
    }
    
    private final void saveExpense() {
    }
    
    private final boolean validateInput(java.lang.String date, java.lang.String startTime, java.lang.String endTime, java.lang.String description, java.lang.String category, java.lang.String amountText) {
        return false;
    }
    
    private final java.time.LocalDateTime parseDateTime(java.lang.String dateString, java.lang.String timeString) {
        return null;
    }
    
    private final java.time.LocalDate parseDate(java.lang.String dateString) {
        return null;
    }
    
    private final java.time.LocalTime parseTime(java.lang.String timeString) {
        return null;
    }
    
    private final boolean validateTimeRange(java.time.LocalDateTime startTime, java.time.LocalDateTime endTime) {
        return false;
    }
}

File: AddExpenseActivity.kt
package com.example.budgettingapp

import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.budgettingapp.data.DataManager
import com.example.budgettingapp.data.Expense
import com.example.budgettingapp.databinding.ActivityAddExpenseBinding
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.*

class AddExpenseActivity : AppCompatActivity() {
    private lateinit var binding: ActivityAddExpenseBinding
    private lateinit var dataManager: DataManager
    
    // Constants for date and time formatting
    private val DATE_FORMATTER = DateTimeFormatter.ISO_LOCAL_DATE
    private val TIME_FORMATTER = DateTimeFormatter.ISO_LOCAL_TIME
    private val CURRENT_DATE = LocalDateTime.now()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAddExpenseBinding.inflate(layoutInflater)
        setContentView(binding.root)

        dataManager = DataManager.getInstance(this)

        setupViews()
    }

    private fun setupViews() {
        // Set current date and time
        binding.etDate.text = DATE_FORMATTER.format(CURRENT_DATE)
        binding.etStartTime.text = TIME_FORMATTER.format(CURRENT_DATE)
        binding.etEndTime.text = TIME_FORMATTER.format(CURRENT_DATE)

        // Set up date picker
        binding.etDate.setOnClickListener {
            showDatePicker()
        }

        // Set up time pickers
        binding.etStartTime.setOnClickListener {
            showTimePicker(binding.etStartTime)
        }

        binding.etEndTime.setOnClickListener {
            showTimePicker(binding.etEndTime)
        }

        // Set up save button
        binding.btnSave.setOnClickListener {
            saveExpense()
        }
    }

    private fun showDatePicker() {
        val calendar = Calendar.getInstance()
        val year = calendar.get(Calendar.YEAR)
        val month = calendar.get(Calendar.MONTH)
        val day = calendar.get(Calendar.DAY_OF_MONTH)

        val datePickerDialog = DatePickerDialog(
            this,
            { _, year, month, dayOfMonth ->
                val selectedDate = "$year-${month + 1}-$dayOfMonth"
                binding.etDate.text = selectedDate
            },
            year,
            month,
            day
        )
        datePickerDialog.show()
    }

    private fun showTimePicker(timeEditText: android.widget.EditText) {
        val calendar = Calendar.getInstance()
        val hour = calendar.get(Calendar.HOUR_OF_DAY)
        val minute = calendar.get(Calendar.MINUTE)

        val timePickerDialog = TimePickerDialog(
            this,
            { _, hourOfDay, minute ->
                val time = String.format("%02d:%02d", hourOfDay, minute)
                timeEditText.text = time
            },
            hour,
            minute,
            true
        )
        timePickerDialog.show()
    }

    private fun saveExpense() {
        val date = binding.etDate.text.toString()
        val startTime = binding.etStartTime.text.toString()
        val endTime = binding.etEndTime.text.toString()
        val description = binding.etDescription.text.toString()
        val category = binding.etCategory.text.toString()
        val amount = binding.etAmount.text.toString().toDoubleOrNull()

        // Validate input
        if (!validateInput(date, startTime, endTime, description, category, amount)) {
            return
        }

        // Parse dates and times
        val parsedDate = parseDate(date)
        val parsedStartTime = parseTime(startTime)
        val parsedEndTime = parseTime(endTime)

        // Validate time range
        if (!validateTimeRange(parsedStartTime, parsedEndTime)) {
            Toast.makeText(this, "End time must be after start time", Toast.LENGTH_SHORT).show()
            return
        }

        val expense = Expense(
            title = description,
            amount = amount,
            category = category,
            date = parsedDate,
            startTime = parsedStartTime,
            endTime = parsedEndTime,
            description = description
        )

        // Save to SharedPreferences using DataManager
        dataManager.saveExpense(expense)
        Toast.makeText(this, "Expense saved successfully", Toast.LENGTH_SHORT).show()
        finish()
    }

    private fun validateInput(
        date: String,
        startTime: String,
        endTime: String,
        description: String,
        category: String,
        amount: Double?
    ): Boolean {
        if (date.isEmpty() || startTime.isEmpty() || endTime.isEmpty() ||
            description.isEmpty() || category.isEmpty() || amount == null) {
            Toast.makeText(this, "Please fill in all fields", Toast.LENGTH_SHORT).show()
            return false
        }

        if (amount < 0) {
            Toast.makeText(this, "Amount must be positive", Toast.LENGTH_SHORT).show()
            return false
        }

        try {
            // Attempt to parse dates and times to ensure they're valid
            parseDate(date)
            parseTime(startTime)
            parseTime(endTime)
        } catch (e: DateTimeParseException) {
            Toast.makeText(this, "Invalid date or time format", Toast.LENGTH_SHORT).show()
            return false
        }

        return true
    }

    private fun parseDate(dateString: String): LocalDateTime {
        return LocalDateTime.parse(dateString, DATE_FORMATTER)
    }

    private fun parseTime(timeString: String): LocalDateTime {
        return LocalDateTime.parse(timeString, TIME_FORMATTER)
    }

    private fun validateTimeRange(startTime: LocalDateTime, endTime: LocalDateTime): Boolean {
        return startTime.isBefore(endTime)
    }
}
